README
Spb - a textual inteface for Slackware.it Package Browser
==========================================================

This is Spb 2.0.


What is it?
-----------

Spb is a textual interface to the Slackware Package Browser. It is a
Perl script that is capable to query and to fetch packages from the
Slackware mirrors. Spb provides also a reusable Perl module (WWW::SPB).


License
-------

Copyright (C), 2006 Davide Angelocola <davide.angelocola@gmail.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
02111-1307 USA.


Web Resources
-------------

The official website is:

  http://dfa.slackware.it/Spb.html


Author
------

Spb was written by Davide Angelocola. See also THANKS for a list of contributors.


Contact Information
-------------------

http://dfa.slackware.it/ContactInfo.html


That's all folks!

-- dfa

(This file was generated from the package metadata on Fri Sep 22 15:45:55 2006)


ChangeLog
2006-09-22  Davide Angelocola  <davide.angelocola@gmail.com>

	* lib/WWW/SPB/versions.pm: new file 

	* configuration file support added
	
	* support for command line options added

	* added 'info' and 'versions' actions

	* library reorganized
	

2006-09-21  Davide Angelocola  <davide.angelocola@gmail.com>

	* merged spb-get, spb-file and spb-package in one script: spb

	* removed URI dependency

2006-09-20  Davide Angelocola  <davide.angelocola@gmail.com>

	* lib/WWW/SPB/version.pm: new file

	* added spb-get

2006-09-18  Davide Angelocola  <davide.angelocola@gmail.com>

	* version 1.2  
	
	* synchronized with new Slackware Package Browser (XHTML 1.1)

	
2006-09-16  Davide Angelocola  <davide.angelocola@gmail.com>

	* updated base url to http://packages.slackware.it/search.php
	



Makefile.PL
#!/usr/bin/env perl

# Makefile.PL -- ExtUtils::MakeMaker input for spb
# Copyright (C) 2006, Davide Angelocola <davide.angelocola@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

use 5.005;
use ExtUtils::MakeMaker;

WriteMakefile
  (
   NAME         => 'spb',
   ABSTRACT     => 'Perform various queries to Slackware Package Browser',
   AUTHOR       => 'Davide Angelocola <davide.angelocola@gmail.com>',
   PREREQ_PM    => { 'LWP', 'HTML::TreeBuilder' },
   VERSION_FROM => 'lib/WWW/SPB.pm',
   depend       => { Makefile => '$(VERSION_FROM)' },
   EXE_FILES    => [ 'bin/spb' ],
   dist         => { COMPRESS => 'gzip -9f' },
);


lib/WWW/SPB.pm
=head1 NAME

 WWW::SPB - fetch and report information from the Slackware Package Browser

=head1 DESCRIPTION

  This simple module is capable to query to the Slackware
  Package Browser in various ways.

=cut

package WWW::SPB;

use strict;
use warnings;

use vars qw( $VERSION );
$VERSION = '2.0';

use Exporter;

our @ISA    = qw(Exporter); 
our @EXPORT = qw(get files packages info versions);

require "WWW/SPB/get.pm";
require "WWW/SPB/search.pm";
require "WWW/SPB/package.pm";
require "WWW/SPB/versions.pm";

=head1 SYNOPSIS

 use WWW::SPB;

 @files = files($pattern [, $slackware_version])
 @package = packages($pattern [, $slackware_version])
 %info = info($package [, $slackware_version])
 get($package, $version)

=head1 WEB RESOURCES

 http://dfa.slackware.it/Spb.html
 http://packages.slackware.it

=head1 SEE ALSO

 spb(1)

=head1 AUTHOR

 Davide Angelocola <davide.angelocola@gmail.com>

=head1 COPYRIGHT

 Copyright (C), 2006 Davide Angelocola <davide.angelocola@gmail.com>

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 MA 02110-1301 USA

=cut



lib/WWW/SPB/get.pm
# get.pm -- download packages
# Copyright (C) 2006, Davide Angelocola <davide.angelocola@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

use strict;
use warnings;

use LWP;

our $total_size;
our $final_data = "";
our $pid;
our $sigint_count = 0;
our $skip_mirror = 0;

sub get {
  my ($package, $version, $mirror) = @_;
  my @mirrors;

  @mirrors = mirrors($package, $version);
  return "package $package ($version) not found" unless @mirrors;

  # prepend user preferred mirror
  unshift @mirrors, $mirror if $mirror;

  $|++;				   # auto-flush

  foreach (@mirrors) {
    $skip_mirror = 0;
    next if $skip_mirror;
    last if $sigint_count == 3;
    next if /^ftp/;		# XXX: FTP support not implemented
    my @url = split /\//, $_;
    $url[0] =~  s/:$//;
    print "trying with $url[2] ($url[0])...\n";

    if ($pid = fork) {
      # parent catches INT and prepare for next download
      local $SIG{INT} =
	sub {
	  $final_data = "";
	  $sigint_count++;
	  $skip_mirror = 1;
	  print "\nCaught keyboard interrupt ($sigint_count/3)\n";
	  kill 9, $pid;
	};

      waitpid($pid, 0);
      return "Ok" if $? == 0;
    } else {
      die "cannot fork: $!" unless defined $pid;
      # child ignores INT and does it's work
      $SIG{INT} = "IGNORE";
      my $res = pretty_mirror($_);
      print "\n";
      if ($res->is_success) {
	open F, ">$package.tgz";
	print F $final_data;
	close F;
	exit 0;
      } else {
	print "Error!\n";
	exit 1;
      }
    }
  }

  return "unable to get '$package' ($version)";
}

# Stolen from:
#   http://hacks.oreilly.com/pub/h/943
sub pretty_mirror {
  my $url = $_[0];
  my $ua = LWP::UserAgent->new;
  $total_size = $ua->head($url)->headers->content_length;
  return $ua->get($url, ':content_cb' => \&get_callback );
}

sub get_callback {
  my ($data, $response, $protocol) = @_;
  $final_data .= $data;
  print progress_bar(length($final_data), $total_size);
}

sub progress_bar {
  my ($got, $total) = @_;
  my $width = 25;
  my $char = '=';
  my $num_width = length $total;

  sprintf "\r|%-${width}s| Got %${num_width}s bytes of %s (%.2f%%)",
    $char x (($width - 1) * $got / $total) . '>',
      $got, $total, 100 * $got /+ $total;
}

1;


lib/WWW/SPB/fetch.pm
# fetch.pm -- download html pages
# Copyright (C) 2006, Davide Angelocola <davide.angelocola@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

use strict;
use warnings;

use LWP;

sub fetch {
  my ($url) = @_;
  my $response;

  $response = LWP::UserAgent->new->get($url);
  return $response->status_line unless $response->is_success;
  return $response->content;
}

1;


lib/WWW/SPB/package.pm
# package.pm -- PB's package.php interface
# Copyright (C) 2006, Davide Angelocola <davide.angelocola@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

use strict;
use warnings;

use HTML::TreeBuilder;

require "WWW/SPB/fetch.pm";

use constant PACKAGE_URL       => 'http://packages.slackware.it/package.php';
use constant PACKAGE_PARAMS    => '?q=%s/%s';

sub package_query {
  return fetch(PACKAGE_URL . sprintf PACKAGE_PARAMS, $_[1], $_[0]);
}

sub mirrors {
  my @mirrors = ();

  my $page = package_query(@_);
  return @mirrors if $page =~ /Sorry/;

  my $root =
    HTML::TreeBuilder->new->parse($page)->find_by_attribute('id', 'pkgmirrors');

  foreach ($root->find_by_tag_name('a')) {
      my $url = $_->attr('href');
      push @mirrors, $url;
    }

  return @mirrors;
}

sub info {
  my $page = package_query(@_);
  my %info = ();

  return %info if $page =~ /Sorry/;

  my $root = HTML::TreeBuilder->new->parse($page);
  my $i = 0;			# to keep insertion order; avoid Tie::IxHash

  foreach ($root->find_by_attribute('id', 'pkghead')->find_by_tag_name('tr')) {
    my @row = $_->content_list;
    $_ = $_->as_text foreach (@row);
    $info{$i++} = sprintf "%-20s %s", $row[0], $row[1];
  }

  # $info{$i} = sprintf "%s\n %s", "description", $root->find_by_attribute('class', 'pkgdescr')->as_text;
  return %info;
}

1;


lib/WWW/SPB/versions.pm
# versions.pm -- fetch version numbers
# Copyright (C) 2006, Davide Angelocola <davide.angelocola@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

use strict;
use warnings;

use HTML::TreeBuilder;

require "WWW/SPB/fetch.pm";

use constant VERSIONS_URL       => 'http://packages.slackware.it/';

sub versions_query {
  return fetch(VERSIONS_URL);
}

sub versions {
  my @versions = ();
  my $root;
  my $v;

  $root = HTML::TreeBuilder->new->parse(versions_query())
    ->find_by_attribute('name', 'v');

  foreach ($root->find_by_tag_name('option')) {
    ($v = $_->as_text) =~ s/slackware-//;
    push @versions, $v;
  }

  return @versions;
}

1;


lib/WWW/SPB/search.pm
# search.pm -- PB's search.php interface
# Copyright (C) 2006, Davide Angelocola <davide.angelocola@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

use strict;
use warnings;

use HTML::TreeBuilder;

require "WWW/SPB/fetch.pm";

use constant SEARCH_URL     => 'http://packages.slackware.it/search.php';
use constant SEARCH_PARAMS  => '?v=%s&t=%s&q=%s';
use constant SEARCH_PACKAGE => 1;
use constant SEARCH_FILE    => 2;

sub search_query {
  my ($item, $version, $type, $index) = @_;
  my $params;

  $params = sprintf(SEARCH_PARAMS, $version, $type,  $item);
  $params .= "&p=$index" if defined $index;
  return fetch(SEARCH_URL . $params);
}

sub get_packages {
  my @packages;

  foreach ($_[0]->find_by_attribute('class', 'result')) {
    my ($name, $category) = split /in /,
      $_->find_by_attribute('class', 'pkgtitle')->as_text;
    push @packages, "$category/$name";
  }

  return @packages;
}

sub get_all_packages {
  my $page;
  my $root;
  my @packages;

  $page = search_query(@_);
  return "spb: unknown Slackware version '$_[1]'"
    if $page =~ /no version/;

  $root = HTML::TreeBuilder->new->parse($page);
  @packages = get_packages($root);

  # get other pages, if needed...
  my $last = 1;
  foreach ($root->find_by_tag_name('select')) {
    next if $_->attr('name') ne 'p';
    my @options = $_->find_by_tag_name('option');
    $last = $options[-1]->as_text if $#options > 1;
    last;
  }

  for (my $i = 2; $i <= $last; $i++) {
    push @packages, 
      get_packages(HTML::TreeBuilder->new->parse(search_query(@_, $i)));
  }

  return @packages;
}

sub files {
  return get_all_packages(@_, SEARCH_FILE);
}

sub packages {
  return get_all_packages(@_, SEARCH_PACKAGE);
}

1;
